### PR TITLE
Added ability to randomly kill pipelines to the constellation.

### PR DESCRIPTION
**Pull request by asajeffrey**
Added flags:
- `--random-pipeline-closure-probability`: probability of each event triggering a forced exit of a randomly chosen pipeline.
- `--random-pipeline-closure-seed`: seed to use for the RNG (to remove a source of intermittent failure).

These are designed for testing the hardness of the constellation.

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/10179)

### CODE REVIEW

**Comment by asajeffrey** [comment]
Things that we need to make sure about this PR:
1. If `--random-pipeline-closure-probability` isn't set, the PR changes nothing.
2. There's no security implications, even if an attacker manages to get servo into zombie apocalypse mode.
I think I managed both of these, but you never know...

**Comment by asajeffrey** [comment]
Argh, forgot to run test-tidy.

**Comment by asajeffrey** [comment]
There is a constellation hardening metabug: issue #10124.

**Comment by emilio** [comment]
Looks good to me!

I left a comment about the `close_random_pipeline` function, but I could tolerate landing it as-is if you think being clearer there is unnecessary, so feel free to r=me if that's the case :)

-S-awaiting-review +S-awaiting-answer 

---

Reviewed 4 of 6 files at r1, 1 of 2 files at r2.
Review status: 5 of 6 files reviewed at latest revision, 1 unresolved discussion.

---

*[components/compositing/constellation.rs, line 370 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10179#-KDiJ6j71JL_O1syCEef:-KDiJ6j71JL_O1syCEeg:-1273323318) ([raw file](https://github.com/servo/servo/blob/e661fa3bf51a83f51f8046f84c022ca3d21f5b56/components/compositing/constellation.rs#L370)):*
I'd include in this comment something like "if `--random-pipeline-closure-probability` was specified at startup" or similar, just to make it clear when scanning the constellation code (in order to not having to look ~1000 lines below).

Renaming `close_random_pipeline` to `maybe_close_random_pipeline`, or even changing the signature to `close_random_pipeline(&mut self, rng: &mut StdRng, probability: f32)`, so it would be obvious from here that this will do nothing unless `random_pipeline_closure` is set would help.

These are nits though, and I won't blame you for landing this as-is :P

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/10179#-:-KDiLEV9Edays_NAkB1l:-855846672)*

**Comment by asajeffrey** [comment]
Review status: 5 of 6 files reviewed at latest revision, 1 unresolved discussion.

---

*[components/compositing/constellation.rs, line 370 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10179#-KDiJ6j71JL_O1syCEef:-KDiRwAr5HK14sI9IXg8:1179119842) ([raw file](https://github.com/servo/servo/blob/e661fa3bf51a83f51f8046f84c022ca3d21f5b56/components/compositing/constellation.rs#L370)):*
Added some comments, and renamed `close_random_pipeline` to `maybe_close_random_pipeline`. Thanks for the review!

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/10179)*

File: components/compositing/constellation.rs [hunk 1]
```diff
@@ -347,6 +352,11 @@ impl<LTF: LayoutThreadFactory, STF: ScriptThreadFactory> Constellation<LTF, STF>
                 child_processes: Vec::new(),
                 document_states: HashMap::new(),
                 webrender_api_sender: state.webrender_api_sender,
+                random_pipeline_closure: opts::get().random_pipeline_closure_probability.map(|prob| {
+                    let seed = opts::get().random_pipeline_closure_seed.unwrap_or_else(random);
```
Comment (mbrubeck) [inline components/compositing/constellation.rs]: When it a random seed, perhaps it should log the seed.

Comment (asajeffrey) [inline components/compositing/constellation.rs]: Done. Should we log using `debug!` or something else?

Comment (mbrubeck) [inline components/compositing/constellation.rs]: `debug!` is compiled out in release builds of Servo, so use `info!` if you want to use this feature in release.

Comment (asajeffrey) [inline components/compositing/constellation.rs]: Will do.

**Comment by emilio** [comment]
Looks good to me, r=me with that typo fixed and the commits squashed :)

-S-awaiting-answer -S-awaiting-answer +S-needs-code-changes +S-needs-squash 

---

Reviewed 1 of 2 files at r2, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[components/compositing/constellation.rs, line 370 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10179#-KDiJ6j71JL_O1syCEef:-KDj6ZYxLQVdF4lqDyj4:450166850) ([raw file](https://github.com/servo/servo/blob/e661fa3bf51a83f51f8046f84c022ca3d21f5b56/components/compositing/constellation.rs#L370)):*
nit: `s/pipelne/pipeline`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/10179#-:-KDj6hqrGEl_fQIpk9K8:549934878)*

**Comment by asajeffrey** [comment]
Squashed, typo fixed.

---

Review status: 5 of 6 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/10179#-:-KDjt4HYvSDJt9LIkvYL:290882623)*

**Comment by emilio** [comment]
Need to update the other lockfiles.

**Comment by asajeffrey** [comment]
Grr, I'll do this tomorrow.

**Comment by asajeffrey** [comment]
OK, fixed Cargo.lock files, Let's see if this works...

@bors-servo try

**Comment by asajeffrey** [comment]
@bors-servo: retry
* #9818

**Pull request merged**